<a href="https://colab.research.google.com/github/dev02chandan/News_Classification/blob/main/News_Classification_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Objective

This project aims at building a model which classifies whether the news given is genuine news or false news by using Natural Language Processing 

##Studying the Data

In [ ]:
import nltk

In [ ]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
import pandas as pd

In [ ]:
fake = pd.read_csv("Fake-211023-185413.csv.xls")
genuine = pd.read_csv("True-211023-185340.csv.xls")

In [ ]:
display(fake.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23481 entries, 0 to 23480
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    23481 non-null  object
 1   text     23481 non-null  object
 2   subject  23481 non-null  object
 3   date     23481 non-null  object
dtypes: object(4)
memory usage: 733.9+ KB


None

In [ ]:
display(genuine.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    21417 non-null  object
 1   text     21417 non-null  object
 2   subject  21417 non-null  object
 3   date     21417 non-null  object
dtypes: object(4)
memory usage: 669.4+ KB


None

In [ ]:
display(genuine.head(10))


title  ...                date
0  As U.S. budget fight looms, Republicans flip t...  ...  December 31, 2017 
1  U.S. military to accept transgender recruits o...  ...  December 29, 2017 
2  Senior U.S. Republican senator: 'Let Mr. Muell...  ...  December 31, 2017 
3  FBI Russia probe helped by Australian diplomat...  ...  December 30, 2017 
4  Trump wants Postal Service to charge 'much mor...  ...  December 29, 2017 
5  White House, Congress prepare for talks on spe...  ...  December 29, 2017 
6  Trump says Russia probe will be fair, but time...  ...  December 29, 2017 
7  Factbox: Trump on Twitter (Dec 29) - Approval ...  ...  December 29, 2017 
8         Trump on Twitter (Dec 28) - Global Warming  ...  December 29, 2017 
9  Alabama official to certify Senator-elect Jone...  ...  December 28, 2017 

[10 rows x 4 columns]

In [ ]:
display(fake.head(10))

title  ...               date
0   Donald Trump Sends Out Embarrassing New Year’...  ...  December 31, 2017
1   Drunk Bragging Trump Staffer Started Russian ...  ...  December 31, 2017
2   Sheriff David Clarke Becomes An Internet Joke...  ...  December 30, 2017
3   Trump Is So Obsessed He Even Has Obama’s Name...  ...  December 29, 2017
4   Pope Francis Just Called Out Donald Trump Dur...  ...  December 25, 2017
5   Racist Alabama Cops Brutalize Black Boy While...  ...  December 25, 2017
6   Fresh Off The Golf Course, Trump Lashes Out A...  ...  December 23, 2017
7   Trump Said Some INSANELY Racist Stuff Inside ...  ...  December 23, 2017
8   Former CIA Director Slams Trump Over UN Bully...  ...  December 22, 2017
9   WATCH: Brand-New Pro-Trump Ad Features So Muc...  ...  December 21, 2017

[10 rows x 4 columns]

In [ ]:
display(fake.subject.value_counts())

News               9050
politics           6841
left-news          4459
Government News    1570
US_News             783
Middle-east         778
Name: subject, dtype: int64

In [ ]:
display(genuine.subject.value_counts())

politicsNews    11272
worldnews       10145
Name: subject, dtype: int64

In [ ]:
fake['target'] = 0
genuine['target'] = 1
# Created a new column - named 'target' which tells us if the news is fake and genuine easily

In [ ]:
display(fake.head(10))

title  ... target
0   Donald Trump Sends Out Embarrassing New Year’...  ...      0
1   Drunk Bragging Trump Staffer Started Russian ...  ...      0
2   Sheriff David Clarke Becomes An Internet Joke...  ...      0
3   Trump Is So Obsessed He Even Has Obama’s Name...  ...      0
4   Pope Francis Just Called Out Donald Trump Dur...  ...      0
5   Racist Alabama Cops Brutalize Black Boy While...  ...      0
6   Fresh Off The Golf Course, Trump Lashes Out A...  ...      0
7   Trump Said Some INSANELY Racist Stuff Inside ...  ...      0
8   Former CIA Director Slams Trump Over UN Bully...  ...      0
9   WATCH: Brand-New Pro-Trump Ad Features So Muc...  ...      0

[10 rows x 5 columns]

In [ ]:
data = pd.concat([fake, genuine], axis = 0)

In [ ]:
data = data.reset_index(drop=True)

In [ ]:
data = data.drop(['subject','date','title'] , axis = 1)
# Dropping the columns that are not required

In [ ]:
print(data.columns) # we will analyze with text and target only

Index(['text', 'target'], dtype='object')


## Steps Involved: 
1. Tokenization 
2. Stemming 
3. Stopword removal
4. Splitting data
4. Vectorization
5. TF-IDF Calculation
6. Logistic Regression


### 1. Tokenization

This process divides a large piece of continuous text into distinct units or tokens basically this process is often known as **Tokenization**.

In [ ]:
from nltk.tokenize import word_tokenize

In [ ]:
data['text'] = data['text'].apply(word_tokenize)

In [ ]:
print(data.head(10))

                                                text  target
0  [Donald, Trump, just, couldn, t, wish, all, Am...       0
1  [House, Intelligence, Committee, Chairman, Dev...       0
2  [On, Friday, ,, it, was, revealed, that, forme...       0
3  [On, Christmas, day, ,, Donald, Trump, announc...       0
4  [Pope, Francis, used, his, annual, Christmas, ...       0
5  [The, number, of, cases, of, cops, brutalizing...       0
6  [Donald, Trump, spent, a, good, portion, of, h...       0
7  [In, the, wake, of, yet, another, court, decis...       0
8  [Many, people, have, raised, the, alarm, regar...       0
9  [Just, when, you, might, have, thought, we, d,...       0


### Stemming


This is the idea of removing the suffix of a word and reducing different forms of a word to a core root.

eg. waiting , waited , waits -> wait

In [ ]:
from nltk.stem.snowball import SnowballStemmer
porter = SnowballStemmer("english")

In [ ]:
def stem_it(text):
  return[porter.stem(word) for word in text]

In [ ]:
data['text'] = data['text'].apply(stem_it)

In [ ]:
print(data.head(10))

                                                text  target
0  [donald, trump, just, couldn, t, wish, all, am...       0
1  [hous, intellig, committe, chairman, devin, nu...       0
2  [on, friday, ,, it, was, reveal, that, former,...       0
3  [on, christma, day, ,, donald, trump, announc,...       0
4  [pope, franci, use, his, annual, christma, day...       0
5  [the, number, of, case, of, cop, brutal, and, ...       0
6  [donald, trump, spent, a, good, portion, of, h...       0
7  [in, the, wake, of, yet, anoth, court, decis, ...       0
8  [mani, peopl, have, rais, the, alarm, regard, ...       0
9  [just, when, you, might, have, thought, we, d,...       0


### Stopword Removal

A stop word is a commonly used word (such as “the”, “a”, “an”, “in”) that a search engine has been programmed to ignore.

In [ ]:
def stop_it(text):
  dt = [word for word in text if len(word)>2]
  return dt

In [ ]:
data['text'] = data['text'].apply(stop_it)

In [ ]:
print(data.head(10))

                                                text  target
0  [donald, trump, just, couldn, wish, all, ameri...       0
1  [hous, intellig, committe, chairman, devin, nu...       0
2  [friday, was, reveal, that, former, milwauke, ...       0
3  [christma, day, donald, trump, announc, that, ...       0
4  [pope, franci, use, his, annual, christma, day...       0
5  [the, number, case, cop, brutal, and, kill, pe...       0
6  [donald, trump, spent, good, portion, his, day...       0
7  [the, wake, yet, anoth, court, decis, that, de...       0
8  [mani, peopl, have, rais, the, alarm, regard, ...       0
9  [just, when, you, might, have, thought, get, b...       0


In [ ]:
data['text'] = data['text'].apply(' '.join)

### Splitting of data
Testing and Training set

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(data['text'], data['target'], test_size=0.25)
display(X_train.head())
print('\n')
display(y_train.head())

33940    washington reuter ohio governor john kasich wh...
17445    the nuclear industri necess for energi product...
13056    oop 2008 bill clinton said the same thing make...
21293    this some pretti surreal stuff the four year s...
6997     the boston globe will publish satir front page...
Name: text, dtype: object

33940    1
17445    0
13056    0
21293    0
6997     0
Name: target, dtype: int64

###Vectorization

The **Vectorization** is a technique used to convert textual data to numerical format. 

Using **Vectorization**, a matrix is created where each column represents a feature and each row represents an individual review

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
my_tfidf = TfidfVectorizer(max_df=0.7)

tfidf_train = my_tfidf.fit_transform(X_train)
tfidf_test = my_tfidf.transform(X_test)

In [ ]:
print(tfidf_train.shape, tfidf_test.shape,"\n\n", y_train.shape)

(33673, 91878) (11225, 91878) 

 (33673,)


### Logistic Regression

Logistic Regression is a Machine Learning algorithm which is used for the classification problems, it is a predictive analysis algorithm and based on the concept of probability

Logistic function 1 / (1 + exp(-x)) normalizes everything to 
be between 0 and 1. Then we can interpret the number you get as a probability.

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score 

In [ ]:
model_1 = LogisticRegression(max_iter = 900, solver = 'lbfgs')
model_1.fit(tfidf_train, y_train)
pred_1 = model_1.predict(tfidf_test)
cr1 = accuracy_score(y_test, pred_1)
print(cr1*100,"%")

98.93095768374165 %


### Passive Agressive Classifier

**Passive**: If the prediction is correct, keep the model and do not make any changes. i.e., the data in the example is not enough to cause any changes in the model. 

**Aggressive**: If the prediction is incorrect, make changes to the model. i.e., some change to the model may correct it.

In [ ]:
from sklearn.linear_model import PassiveAggressiveClassifier

model = PassiveAggressiveClassifier(max_iter=50)
model.fit(tfidf_train, y_train)

PassiveAggressiveClassifier(max_iter=50)

In [ ]:
y_pred = model.predict(tfidf_test)
acc_score = accuracy_score(y_test, y_pred)
print("The Accuracy of the prediction is : ", acc_score*100)

The Accuracy of the prediction is :  99.72383073496658


## Conclusion



In conclusion we can say that we have successfully created a model whether a news is genuine or fake using Natural Language Processing with 99.724 % Accuracy